### PR TITLE
Install npm as per engines.npm

### PR DESCRIPTION
**Pull request by ofrobots**
**DO NOT LAND YET**. This should land after the beta runtime builder becomes the default.

If the package.json specifies an engines.npm field, fetch and install that version of npm.

### CODE REVIEW

File: builder/steps/gen-dockerfile/contents/test/detect_setup_test.ts [hunk 1]
```diff
@@ -14,6 +14,11 @@
  * limitations under the License.
  */
 
+// Describe the signature of function exported by the `shell-escape` module
+// so the compiler knows what types to expect and return.  This is needed
+// since the `shell-escape` module does not have type definitions available.
+const shellEscape: (args: string[]) => string = require('shell-escape');
```
Comment (matthewloring) [inline builder/steps/gen-dockerfile/contents/test/detect_setup_test.ts]: I'm curious what the recommended practice is for writing type signatures for untyped modules. In the case of a module that exports a single function this seems fine but for more complex or widely required modules, inlining type definitions at each require point seems like it would be a lot of duplicated code. Is there a way to specify project local type definitions for external modules?

File: builder/steps/gen-dockerfile/contents/gulpfile.js [hunk 2]
```diff
@@ -94,7 +94,8 @@ gulp.task('test.compile', ['compile'], function() {
 });
 
 gulp.task('test.unit', ['test.compile'], function() {
-  return gulp.src([outDir + '/test/**/*.js']).pipe(mocha({reporter: 'spec'}));
+  return gulp.src([outDir + '/test/**/*.js'])
+      .pipe(mocha({reporter: 'spec', require: 'source-map-support/register'}));
```
Comment (matthewloring) [inline builder/steps/gen-dockerfile/contents/gulpfile.js]: What prompted this change? It doesn't seem related to `engines.npm`?

File: builder/steps/gen-dockerfile/contents/src/detect_setup.ts [hunk 3]
```diff
@@ -133,14 +139,11 @@ export async function detectSetup(
...
-      nodeVersion = packageJson.engines.node;
-    } else {
-      nodeVersion = undefined;
-      warn(
-          'node.js checker: ignoring invalid "engines" field in ' +
-          'package.json');
+    // Check if the user has specified specific versions of node or npm in the
+    // package.json.
+    if (packageJson.engines) {
+      nodeVersion = packageJson.engines.node || undefined;
```
Comment (matthewloring) [inline builder/steps/gen-dockerfile/contents/src/detect_setup.ts]: What cases are the `|| undefined` intended to catch? If `packageJson.engines.node` or `packageJson.engines.npm` are falsey but not `undefined` should we be printing a more detailed error message?

File: builder/steps/gen-dockerfile/contents/gulpfile.js (hunk 2 above)
Comment (ofrobots) [inline builder/steps/gen-dockerfile/contents/gulpfile.js]: It is not related indeed. It makes sure that mocha stack traces show TS rather than JS stacks.

Comment (ofrobots) [inline builder/steps/gen-dockerfile/contents/gulpfile.js]: Will put this into a separate PR.

File: builder/steps/gen-dockerfile/contents/src/detect_setup.ts (hunk 3 above)
Comment (ofrobots) [inline builder/steps/gen-dockerfile/contents/src/detect_setup.ts]: Done.

**Comment by DominicKramer** [comment]
This PR has been replaced with PR #127.
